### DISCUSSION TITLE
ファイルの読み込み・書き込みの遅延の可能性について

### DISCUSSION THREAD

**Discussion by mkoki-hrk** [nig-sc/nigsc_homepage2#417] [Q&A] [answered]
いつも大変お世話になっております。

昨年12月末頃から、ファイルの読み書きが遅くなったと感じています。
ls (ディレクトリ中のファイルが数個程度でも), vi, man等のプログラムで体感的に応答が遅くなることがありました。

それを異常と感じたきっかけは、bashの起動に時間がかかる（30秒程度）ようになったことでした。
\~/.bashrcや\~/.bash_profileの処理のどこに時間がかかっているのか調べると、私がminicondaをインストールした際に書き込まれた、\~/.bashrc中のconda_initializeの処理`__conda_setup="$('/home/username/mambaforge/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"`にありました。

そこで、`$ strace '/home/username/mambaforge/bin/conda' 'shell.bash' 'hook'`を実行し、発行されたシステムコールと処理時間を確認すると、`openat()`や`close()`で処理が滞る場合があると判明しました。
このため、ファイルI/Oの遅延を疑うようになりました。

以下、試みに空ディレクトリ中に空ファイルを作成した際の、straceからの出力です。
`openat()`に1.5秒程度時間がかかっているとわかりました。

```
[username@at137 ~]$ mkdir tmp && cd tmp
[username@at137 ~/tmp]$ strace -T touch hoge
execve("/usr/bin/touch", ["touch", "hoge"], 0x7ffcea4a1b30 /* 41 vars */) = 0 <0.000222>
(略)
openat(AT_FDCWD, "hoge", O_WRONLY|O_CREAT|O_NOCTTY|O_NONBLOCK, 0666) = 3 <1.495968>
dup2(3, 0)                              = 0 <0.000008>
close(3)                                = 0 <0.000008>
utimensat(0, NULL, NULL, 0)             = 0 <0.007475>
close(0)                                = 0 <0.002407>
close(1)                                = 0 <0.000006>
close(2)                                = 0 <0.000006>
exit_group(0)                           = ?
+++ exited with 0 +++
```
（ユーザー名をusernameに置き換えています）

[Lustreファイルシステムの解説をされたページ](https://sc.ddbj.nig.ac.jp/general_analysis_division/ga_lustre)に、大量のファイルの入ったディレクトリやメタデータの操作ではパフォーマンスが低下するとの言及がありましたが、Lustreファイルシステムではこれはあり得る（正常な）状態でしょうか？異常とすれば、原因はどこにあるでしょうか？

解決方法について、ご教示いただけますと幸いです。
お手数ですが、どうぞよろしくお願いいたします。

**Comment by sc-fs** [comment, marked as answer]
ご迷惑をお掛けしております。
大幅な遅延の原因は、
大容量高速ストレージ（Lustre7）の特定領域(OST)の使用率が高かったため、他の領域との使用率に大きく乖離が発生し、
その影響で、本来並列ファイルシステムとしての大容量高速ストレージ（Lustre）が持つ性能が十分に発揮できていなかったためです。

対応として、使用率が高かった特定領域(OST)を他の領域と同じになるように対処しました。

現在は、本来の状態でアクセス出来ております。

**Reply by mkoki-hrk** [reply]
ご対応いただき、ありがとうございました。
遅延の解消が確認できました。大変助かります。
どうぞよろしくお願いいたします。
